File: 22-generate-parentheses/generate-parentheses.py
class Solution:
    def generateParenthesis(self, n: int) -> List[str]:
        if n == 0:
            return ['']
        result = []
        
        def backtrack(temp, open_p, close_p):
            if len(temp) == 2 * n:
                result.append(temp)
                return
            if open_p < n:
                backtrack(temp + "(", open_p + 1, close_p)
            if close_p < open_p:
                backtrack(temp + ")", open_p, close_p + 1)
            
        backtrack("", 0, 0)
        return result

File: 22-generate-parentheses/generate-parentheses.java
class Solution {
    private List<String> result = new ArrayList<>();
    private int num;
    private void backtrack(StringBuilder temp, int open, int close){
        if(temp.length() == 2*num){
            result.add(temp.toString());
            return;
        }

        if(open < num){
            temp.append('(');
            backtrack(temp,open+1, close);
            temp.deleteCharAt(temp.length()-1);
        }
        if(close < open){
            temp.append(')');
            backtrack(temp,open, close+1);
            temp.deleteCharAt(temp.length()-1);
        }
    }

    public List<String> generateParenthesis(int n) {
        num = n;
        backtrack(new StringBuilder(), 0, 0);
        return result;
    }
}